{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ab1f9d0b",
   "metadata": {},
   "source": [
    "# Imputation and DGE analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72dbd17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import os, sys, glob, re, math, pickle\n",
    "import phate, scprep, magic\n",
    "import graphtools as gt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import time,random,datetime\n",
    "import networkx as nx\n",
    "from sklearn import metrics\n",
    "from sklearn import model_selection\n",
    "from scipy import sparse\n",
    "from scipy.stats import mannwhitneyu, tiecorrect, rankdata\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import scanpy as sc\n",
    "from sklearn.dummy import DummyClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.cluster import SpectralClustering, OPTICS, cluster_optics_dbscan, AgglomerativeClustering\n",
    "from bbknn import bbknn\n",
    "import warnings\n",
    "%matplotlib inline\n",
    "%load_ext memory_profiler\n",
    "import logging\n",
    "\n",
    "# settings\n",
    "plt.rc('font', size = 8)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "sc.set_figure_params(dpi=300,dpi_save=600,\n",
    "                     frameon=False,\n",
    "                     fontsize=8)\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sc.settings.verbosity=2\n",
    "sc._settings.ScanpyConfig.n_jobs=-1\n",
    "\n",
    "# reproducibility\n",
    "rs = np.random.seed(42)\n",
    "\n",
    "# utils\n",
    "def mwu(X,Y,gene_names,correction=None,debug=False) :\n",
    "    '''\n",
    "    Benjamini-Hochberg correction implemented. Can change to Bonferonni\n",
    "    gene_names (list)\n",
    "    if X,Y single gene expression array, input x.reshape(-1,1), y.reshape(-1,1)\n",
    "    NOTE: get zeros sometimes because difference (p-value is so small)\n",
    "    '''\n",
    "    p=pd.DataFrame()\n",
    "    print('Mann-Whitney U w/Benjamini/Hochberg correction\\n')\n",
    "    start = time.time()\n",
    "    for i,g in enumerate(gene_names) :\n",
    "        if i==np.round(np.quantile(np.arange(len(gene_names)),0.25)) :\n",
    "            print('... 25% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.5)) :\n",
    "            print('... 50% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.75)) :\n",
    "            print('... 75% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        p.loc[i,'Gene']=g\n",
    "        if (tiecorrect(rankdata(np.concatenate((np.asarray(X[:,i]),np.asarray(Y[:,i])))))==0) :\n",
    "            if debug :\n",
    "                print('P-value not calculable for {}'.format(g))\n",
    "            p.loc[i,'pval']=np.nan\n",
    "        else :\n",
    "            _,p.loc[i,'pval']=mannwhitneyu(X[:,i],Y[:,i]) # continuity correction is True\n",
    "    print('... mwu computed in {:.2f}-s\\n'.format(time.time() - start))\n",
    "    # ignore NaNs, since can't do a comparison on these (change numbers for correction)\n",
    "    p_corrected = p.loc[p['pval'].notna(),:]\n",
    "    new_pvals = multipletests(p_corrected['pval'],method='fdr_bh')\n",
    "    p_corrected['pval_corrected'] = new_pvals[1]\n",
    "    return p_corrected\n",
    "\n",
    "def log2aveFC(X,Y,gene_names,AnnData=None) :\n",
    "    '''not sensitivity to directionality due to subtraction\n",
    "    X and Y full arrays, subsetting performed here\n",
    "    `gene_names` (list): reduced list of genes to calc\n",
    "    `adata` (sc.AnnData): to calculate reduced list. NOTE: assumes X,Y drawn from adata.var_names\n",
    "    '''\n",
    "    if not AnnData is None :\n",
    "        g_idx = [i for i,g in enumerate(AnnData.var_names) if g in gene_names]\n",
    "        fc=pd.DataFrame({'Gene':AnnData.var_names[g_idx],\n",
    "                         'log2FC':np.log2(X[:,g_idx].mean(axis=0)) - np.log2(Y[:,g_idx].mean(axis=0))}) # returns NaN if negative value \n",
    "    else :\n",
    "        fc=pd.DataFrame({'Gene':gene_names,\n",
    "                         'log2FC':np.log2(X.mean(axis=0)) - np.log2(Y.mean(axis=0))})\n",
    "    return fc\n",
    "\n",
    "\n",
    "# fps\n",
    "dfp = '/vast/palmer/pi/lim_janghoo/cl2292/'\n",
    "pfp = '/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/results/'\n",
    "pdfp = '/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/data/'\n",
    "sc.settings.figdir = pfp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cc72e6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Imputation by genotype & timepoint\n",
    "\n",
    "wt30 = adata[(adata.obs['genotype']=='SCA1-fl/+'), :]\n",
    "mut30 = adata[(adata.obs['genotype']=='SCA1-fl/NG2-Cre'), :]\n",
    "\n",
    "#k=45, t=3\n",
    "print('Starting imputation for {}\\n'.format('30wk SCA1-fl/+'))\n",
    "tic = time.time()\n",
    "\n",
    "\n",
    "wt30.obs['value'] = 0\n",
    "sc.pp.pca(wt30)\n",
    "sc.pp.neighbors(wt30, n_pcs=45)\n",
    "\n",
    "\n",
    "# MAGIC\n",
    "G = gt.Graph(data=wt30.obsp['connectivities']+sparse.diags([1]*wt30.shape[0],format='csr'),\n",
    "             precomputed='adjacency',\n",
    "             use_pygsp=True)\n",
    "G.knn_max = None\n",
    "\n",
    "magic_op=magic.MAGIC().fit(X=wt30.X,graph=G) # running fit_transform produces wrong shape\n",
    "wt30.layers['imputed']=magic_op.transform(wt30.X,genes='all_genes')\n",
    "\n",
    "print('\\n  imputation in {:.2f}-min'.format((time.time() - tic)/60))\n",
    "\n",
    "\n",
    "print('\\n Starting imputation for {}\\n'.format('30wk SCA1-fl/NG2-Cre'))\n",
    "tic = time.time()\n",
    "\n",
    "mut30.obs['value'] = 0\n",
    "sc.pp.pca(mut30)\n",
    "sc.pp.neighbors(mut30, n_pcs=45)\n",
    "\n",
    "# MAGIC\n",
    "G = gt.Graph(data=mut30.obsp['connectivities']+sparse.diags([1]*mut30.shape[0],format='csr'),\n",
    "             precomputed='adjacency',\n",
    "             use_pygsp=True)\n",
    "G.knn_max = None\n",
    "\n",
    "magic_op=magic.MAGIC().fit(X=mut30.X,graph=G) # running fit_transform produces wrong shape\n",
    "mut30.layers['imputed']=magic_op.transform(mut30.X,genes='all_genes')\n",
    "\n",
    "print('\\n  imputation in {:.2f}-min'.format((time.time() - tic)/60))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "705d8e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata =wt30.concatenate([mut30,] ,batch_key='concat')\n",
    "\n",
    "# save data objects\n",
    "adata.write(os.path.join(pdfp,'250414_OL-SCA1-cKI_imp.h5ad'))\n",
    "print('saved @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f8dbee",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EMD; IMP\n",
    "wt = adata[(adata.obs['genotype']=='SCA1-fl/+'), :]\n",
    "mut = adata[(adata.obs['genotype']=='SCA1-fl/NG2-Cre'), :]\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='ctype'\n",
    "    fname = 'SCA1-flwCre vs SCA1-flwoCre' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['30wk'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in wt.obs[group].unique():\n",
    "            start = time.time()\n",
    "            print('\\n{}, SCA1 vs WT'.format(i))\n",
    "            print('----')\n",
    "            X = wt[(wt.obs[group]==i)&(wt.obs['timepoint']==t), :].layers['imputed']\n",
    "            Y = mut[(mut.obs[group]==i)&(mut.obs['timepoint']==t), :].layers['imputed']\n",
    "            \n",
    "\n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(Y,X,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(Y,X, wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = pd.concat([dge,dt], ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dge.to_csv(os.path.join(pfp,'250414_dge_'+fname+'.csv'),index=False)\n",
    "#     dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "#     dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "#     dgeup.to_csv(os.path.join(pfp,'240723_dge_'+fname+'_SCA1flwCre_up.csv'),index=False)\n",
    "#     dgedown.to_csv(os.path.join(pfp,'240723_dge_'+fname+'_SCA1flwCre_down.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "068f9172",
   "metadata": {},
   "outputs": [],
   "source": [
    "## DEG numbers\n",
    "\n",
    "num_dge=pd.DataFrame()\n",
    "\n",
    "dge = pd.read_csv('/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/results/250414_dge_SCA1-flwCre vs SCA1-flwoCre.csv')\n",
    "\n",
    "downsig = dge.loc[(dge['emd']<-0.1)&(dge['pval_corrected']<0.01),:]\n",
    "upsig = dge.loc[(dge['emd']>0.1)&(dge['pval_corrected']<0.01),:]\n",
    "\n",
    "for t in ['30wk']:\n",
    "    for c in adata.obs['ctype'].unique():\n",
    "        down = downsig.loc[(downsig['Cell type']==c)&(downsig['timepoint']==t),:]\n",
    "        up = upsig.loc[(upsig['Cell type']==c)&(upsig['timepoint']==t),:]\n",
    "        de = pd.DataFrame({'Cell type':[str(c)], 'timepoint':[str(t)], 'downregulated':[len(down['Gene'])], 'upregulated':[len(up['Gene'])]})\n",
    "        num_dge = pd.concat([num_dge,de], ignore_index=True)\n",
    "#         down['Gene'].to_csv(os.path.join(pfp,'231006_'+t+'_'+c+'_down_ASOvsWT.csv'),index=False)\n",
    "#         up['Gene'].to_csv(os.path.join(pfp,'231006_'+t+'_'+c+'_up_ASOvsWT.csv'),index=False)\n",
    "num_dge.to_csv(os.path.join(pfp,'250414_number of DEG.csv'),index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86d4605e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, glob, re, math, pickle\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import csv\n",
    "import gseapy\n",
    "pfp = '/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/results/'\n",
    "\n",
    "\n",
    "# settings\n",
    "plt.rc('font', size = 9)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "def filterdeg(df, ctype, timepoint=None):\n",
    "    if 'Cell type' not in df.columns or 'Gene' not in df.columns:\n",
    "        raise ValueError(\"DataFrame must contain 'Cell type' and 'Gene' columns.\")\n",
    "    \n",
    "    if timepoint is None:\n",
    "        filtered_df = df.loc[df['Cell type'] == ctype, :]\n",
    "    else:\n",
    "        if 'timepoint' not in df.columns:\n",
    "            raise ValueError(\"DataFrame must contain 'timepoint' column for filtering by timepoint.\")\n",
    "        filtered_df = df.loc[(df['Cell type'] == ctype) & (df['timepoint'] == timepoint), :]\n",
    "\n",
    "    return filtered_df['Gene'].to_list()\n",
    "\n",
    "def enrichr(genes, title = 'Title',geneset = 'GO_Biological_Process_2023', save = None):\n",
    "    genes = genes\n",
    "    res=gseapy.enrichr(gene_list=genes, organism = 'Mouse', gene_sets = geneset, cutoff=0.05)\n",
    "    df = res.res2d[res.res2d['Adjusted P-value']<0.05]\n",
    "    df = df.sort_values(by=['Combined Score'], ascending = False).head(10)\n",
    "    df['-log10(Adjusted P-value)'] = df['Adjusted P-value'].apply(np.log10)*-1\n",
    "    df['Term'] = df['Term'].apply(lambda x: x[:-13]) #Remove '(GO:xxxxxxxxx)'\n",
    "\n",
    "    # Data\n",
    "    GO_biological_processes = df['Term'].to_list()\n",
    "    Fold_Enrichment = df['Combined Score'].to_list()\n",
    "    bar_colors = df['-log10(Adjusted P-value)'].to_list()\n",
    "\n",
    "    # Check if bar_colors is empty, and skip plotting if it is\n",
    "    if not bar_colors:\n",
    "        print(\"No significant GO terms to plot.\")\n",
    "        return\n",
    "    \n",
    "    # Create figure and axis\n",
    "    fig, ax = plt.subplots(1,1, figsize = (2,2))\n",
    "\n",
    "    # Create horizontal bars\n",
    "    bars = ax.barh(GO_biological_processes, Fold_Enrichment, color='gray')\n",
    "\n",
    "    # Create a ScalarMappable for color mapping\n",
    "    sm = plt.cm.ScalarMappable(cmap='viridis', norm=plt.Normalize(vmin=min(bar_colors), vmax=max(bar_colors)))\n",
    "\n",
    "    # Convert p-values to colors using the colormap\n",
    "    colors = [sm.to_rgba(p_value) for p_value in bar_colors]\n",
    "\n",
    "    # Set colors for the bars\n",
    "    for bar, color in zip(bars, colors):\n",
    "        bar.set_color(color)\n",
    "\n",
    "    # Create a divider to make room for the colorbar\n",
    "    divider = make_axes_locatable(ax)\n",
    "\n",
    "    # Append an axis for the colorbar on the right side of the main plot\n",
    "    cax = divider.append_axes(\"right\", size=\"5%\", pad=0.05)\n",
    "\n",
    "    # Add a colorbar to the plot\n",
    "    cbar = plt.colorbar(sm, cax=cax)\n",
    "    cbar.set_label('-log10(Adjusted P-value)')\n",
    "\n",
    "    # Set labels and title\n",
    "    ax.set_xlabel('Combined Score')\n",
    "    ax.set_ylabel('GO Biological Process')\n",
    "    ax.set_title(title)\n",
    "\n",
    "    # Adjust x tick labels font size\n",
    "    plt.xticks(fontsize=8)\n",
    "\n",
    "    if save is None:\n",
    "        plt.show()\n",
    "    else:\n",
    "        fig.savefig(os.path.join(pfp,save),dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "\n",
    "def intersection(lst1, lst2):\n",
    "    lst3 = [value for value in lst1 if value in lst2]\n",
    "    return lst3    \n",
    "\n",
    "\n",
    "dge = pd.read_csv('/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/results/250414_dge_SCA1-flwCre vs SCA1-flwoCre.csv')\n",
    "\n",
    "downsig = dge.loc[(dge['emd']<-0.1)&(dge['pval_corrected']<0.01),:]\n",
    "upsig = dge.loc[(dge['emd']>0.1)&(dge['pval_corrected']<0.01),:]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4106aa55",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort by -log(adj_p)\n",
    "def enrichr2(genes, title = 'Title',geneset = 'GO_Biological_Process_2023', save = None):\n",
    "    genes = genes\n",
    "    res=gseapy.enrichr(gene_list=genes, organism = 'Mouse', gene_sets = geneset, cutoff=0.05)\n",
    "    df = res.res2d[res.res2d['Adjusted P-value']<0.05]\n",
    "    df['-log10(Adjusted P-value)'] = df['Adjusted P-value'].apply(np.log10)*-1\n",
    "    df = df.sort_values(by=['-log10(Adjusted P-value)'], ascending = False).head(10)\n",
    "    df['Term'] = df['Term'].apply(lambda x: x[:-13]) #Remove '(GO:xxxxxxxxx)'\n",
    "\n",
    "    # Data\n",
    "    GO_biological_processes = df['Term'].to_list()\n",
    "    Fold_Enrichment = df['-log10(Adjusted P-value)'].to_list()\n",
    "    bar_colors = df['Combined Score'].to_list()\n",
    "\n",
    "    # Check if bar_colors is empty, and skip plotting if it is\n",
    "    if not bar_colors:\n",
    "        print(\"No significant GO terms to plot.\")\n",
    "        return\n",
    "    \n",
    "    # Create figure and axis\n",
    "    fig, ax = plt.subplots(1,1, figsize = (2,2))\n",
    "\n",
    "    # Create horizontal bars\n",
    "    bars = ax.barh(GO_biological_processes, Fold_Enrichment, color='gray')\n",
    "\n",
    "    # Create a ScalarMappable for color mapping\n",
    "    sm = plt.cm.ScalarMappable(cmap='viridis', norm=plt.Normalize(vmin=min(bar_colors), vmax=max(bar_colors)))\n",
    "\n",
    "    # Convert p-values to colors using the colormap\n",
    "    colors = [sm.to_rgba(p_value) for p_value in bar_colors]\n",
    "\n",
    "    # Set colors for the bars\n",
    "    for bar, color in zip(bars, colors):\n",
    "        bar.set_color(color)\n",
    "\n",
    "    # Create a divider to make room for the colorbar\n",
    "    divider = make_axes_locatable(ax)\n",
    "\n",
    "    # Append an axis for the colorbar on the right side of the main plot\n",
    "    cax = divider.append_axes(\"right\", size=\"5%\", pad=0.05)\n",
    "\n",
    "    # Add a colorbar to the plot\n",
    "    cbar = plt.colorbar(sm, cax=cax)\n",
    "    cbar.set_label('Combined Score')\n",
    "\n",
    "    # Set labels and title\n",
    "    ax.set_xlabel('-log10(Adjusted P-value)')\n",
    "    ax.set_ylabel('GO Biological Process')\n",
    "    ax.set_title(title)\n",
    "\n",
    "    # Adjust x tick labels font size\n",
    "    plt.xticks(fontsize=8)\n",
    "\n",
    "    if save is None:\n",
    "        plt.show()\n",
    "    else:\n",
    "        fig.savefig(os.path.join(pfp,save),dpi=300, bbox_inches='tight')\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d729294c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## PC \n",
    "geneset = 'GO_Cellular_Component_2025'\n",
    "\n",
    "c_list = ['PC']\n",
    "for c in c_list:\n",
    "    for t in ['30wk',]:\n",
    "        cKIdown = set(filterdeg(downsig, ctype=c, timepoint=t))\n",
    "        cKIup = set(filterdeg(upsig, ctype=c, timepoint=t))\n",
    "\n",
    "#         enrichr(list(cKIdown), title = c+', '+t+' down cKI',\n",
    "#                 geneset = geneset, \n",
    "#                 save = '250415_enrichR_'+c+'_'+t+'_cKI down_CC.pdf'\n",
    "#                )\n",
    "        \n",
    "        enrichr2(list(cKIup), title = c+', '+t+' up cKI',\n",
    "                geneset = geneset,\n",
    "                save = '250415_enrichR_'+c+'_'+t+'_cKI up_CC_sortbyp.pdf'\n",
    "               )\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "200b83fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## PC \n",
    "geneset = 'GO_Biological_Process_2025'\n",
    "\n",
    "c_list = ['OPC','OL']\n",
    "for c in c_list:\n",
    "    for t in ['30wk',]:\n",
    "        cKIdown = set(filterdeg(downsig, ctype=c, timepoint=t))\n",
    "        cKIup = set(filterdeg(upsig, ctype=c, timepoint=t))\n",
    "\n",
    "        enrichr(list(cKIdown), title = c+', '+t+' down cKI',\n",
    "                geneset = geneset, \n",
    "                save = '250415_enrichR_'+c+'_'+t+'_cKI down_BP.pdf'\n",
    "               )\n",
    "        \n",
    "        enrichr(list(cKIup), title = c+', '+t+' up cKI',\n",
    "                geneset = geneset,\n",
    "                save = '250415_enrichR_'+c+'_'+t+'_cKI up_BP.pdf'\n",
    "               )\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2a12805",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Biological Process sorted by p-value\n",
    "\n",
    "geneset = 'GO_Biological_Process_2025'\n",
    "sort_by = 'p_value'\n",
    "# downsig = cKIdown\n",
    "# upsig = cKIup\n",
    "\n",
    "c_list = ['OL']\n",
    "\n",
    "\n",
    "for t in ['30wk']:\n",
    "    all_results=[]\n",
    "    for c in c_list:\n",
    "        down = set(filterdeg(downsig, ctype=c, timepoint=t))\n",
    "        up = set(filterdeg(upsig, ctype=c, timepoint=t))\n",
    "\n",
    "        res = gseapy.enrichr(gene_list=list(down), organism='mouse', gene_sets=geneset, cutoff=0.05)\n",
    "        df1 = res.res2d[res.res2d['Adjusted P-value'] < 0.05]\n",
    "        res = gseapy.enrichr(gene_list=list(up), organism='mouse', gene_sets=geneset, cutoff=0.05)\n",
    "        df2 = res.res2d[res.res2d['Adjusted P-value'] < 0.05]\n",
    "        \n",
    "        df1['Cell type'] = c\n",
    "        df2 = df2.copy()\n",
    "        df2['Cell type'] = c\n",
    "        df1['Timepoint'] = t\n",
    "        df2 = df2.copy()\n",
    "        df2['Timepoint'] = t\n",
    "        \n",
    "        # Sort by the chosen method\n",
    "        if sort_by == 'combined_score':\n",
    "            df1['Combined Score'] = df1['Combined Score']*-1\n",
    "            df1 = df1.sort_values(by=['Combined Score'], ascending=False).head(5)\n",
    "            df2 = df2.sort_values(by=['Combined Score'], ascending=False).head(5)\n",
    "            df = pd.concat([df1, df2])\n",
    "            df = df.sort_values(by=['Combined Score'], ascending = False)            \n",
    "            color_data = df['-log10(Adjusted P-value)'] = df['Adjusted P-value'].apply(np.log10) * -1\n",
    "        elif sort_by == 'p_value':\n",
    "            \n",
    "            df1['-log10(Adjusted P-value)'] = df1['Adjusted P-value'].apply(np.log10)\n",
    "            df2 = df2.copy()\n",
    "            df2['-log10(Adjusted P-value)'] = df2['Adjusted P-value'].apply(np.log10) * -1\n",
    "\n",
    "            df1 = df1.sort_values(by=['-log10(Adjusted P-value)'], ascending=False).head(10)\n",
    "            df2 = df2.sort_values(by=['-log10(Adjusted P-value)'], ascending=False).head(10)\n",
    "            df = pd.concat([df1, df2])\n",
    "            df = df.sort_values(by=['-log10(Adjusted P-value)'], ascending = False)                \n",
    "        else:\n",
    "            raise ValueError(\"sort_by must be 'combined_score' or 'p_value'\")\n",
    "\n",
    "        all_results.append(df)\n",
    "        \n",
    "    df_combined = pd.concat(all_results)\n",
    "\n",
    "    if sort_by == 'combined_score':          \n",
    "        x_data = df_combined['Combined Score'].to_list()\n",
    "        color_data = df_combined['-log10(Adjusted P-value)'].to_list()\n",
    "        x_label = 'Combined Score'\n",
    "        color_label = '-log10(Adjusted P-value)'\n",
    "        cmap_label = '-log10(Adjusted P-value)'\n",
    "    elif sort_by == 'p_value':              \n",
    "        x_data = df_combined['-log10(Adjusted P-value)'].to_list()\n",
    "        color_data = df_combined['Combined Score'].to_list()\n",
    "        x_label = '-log10(Adjusted P-value)'\n",
    "        color_label = 'Combined Score'\n",
    "        cmap_label = 'Combined Score'\n",
    "    else:\n",
    "        raise ValueError(\"sort_by must be 'combined_score' or 'p_value'\")\n",
    "            \n",
    "\n",
    "    df_combined['Term'] = df_combined['Term'].apply(lambda x: x[:-13])  # Remove '(GO:xxxxxxxxx)'\n",
    "#     df_combined['Term2'] = df_combined['Timepoint']+'_'+df_combined['Cell type']+ ' '+df_combined['Term']\n",
    "    GO_biological_processes = df_combined['Term'].to_list()\n",
    "\n",
    "#     # Check if there's data to plot\n",
    "#     if df.empty:\n",
    "#         print(f\"No significant GO terms to plot for {title}.\")\n",
    "#         return\n",
    "    # Create figure and axis\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(2,2))\n",
    "\n",
    "    # Create horizontal bars\n",
    "    bars = ax.barh(GO_biological_processes, x_data, color='gray')\n",
    "\n",
    "    # Create a ScalarMappable for color mapping\n",
    "    sm = plt.cm.ScalarMappable(cmap='viridis', norm=plt.Normalize(vmin=min(color_data), vmax=max(color_data)))\n",
    "\n",
    "    # Set colors for the bars\n",
    "    colors = [sm.to_rgba(score) for score in color_data]\n",
    "    for bar, color in zip(bars, colors):\n",
    "        bar.set_color(color)\n",
    "\n",
    "        \n",
    "#     # Add labels for cell types\n",
    "#     for bar, cell_type in zip(bars, df_combined['Cell type']):\n",
    "#         ax.text(\n",
    "#             bar.get_width(),  # x-coordinate of the label\n",
    "#             bar.get_y() + bar.get_height() / 2,  # y-coordinate of the label (centered vertically)\n",
    "#             cell_type,  # Label text\n",
    "#             va='center',  # Vertical alignment\n",
    "#             ha='left',  # Horizontal alignment\n",
    "#             fontsize=8,  # Font size for the label\n",
    "#             color='black'  # Text color\n",
    "#         )\n",
    "        \n",
    "        \n",
    "    # Create a divider to make room for the colorbar\n",
    "    divider = make_axes_locatable(ax)\n",
    "\n",
    "    # Append an axis for the colorbar on the right side of the main plot\n",
    "    cax = divider.append_axes(\"right\", size=\"5%\", pad=0.05)\n",
    "\n",
    "    # Add a colorbar to the plot\n",
    "    cbar = plt.colorbar(sm, cax=cax)\n",
    "    cbar.set_label(cmap_label, color = 'black')\n",
    "\n",
    "    # Set labels and title\n",
    "    ax.set_xlabel(x_label, color = 'black')\n",
    "    ax.set_ylabel(geneset, color = 'black')\n",
    "    ax.set_title(t, color = 'black')\n",
    "    ax.grid(False)\n",
    "    \n",
    "    # Customize axes tick labels to be black\n",
    "    ax.tick_params(axis='x', colors='black')\n",
    "    ax.tick_params(axis='y', colors='black')\n",
    "    ax.axvline(x=0, color = 'black', linestyle='-', linewidth=0.5)\n",
    "\n",
    "    # Adjust x tick labels font size\n",
    "    plt.xticks(fontsize=8, color='black')\n",
    "    plt.yticks(color='black')\n",
    "\n",
    "    # Show or save the plot\n",
    "#     fig.savefig(os.path.join(pfp, '250415_cKI_OL_enrichr_'+t+'_pval.pdf'), dpi=300, bbox_inches='tight')\n",
    "    plt.show()\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf0a34b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, glob, re, math, pickle\n",
    "import pandas as pd\n",
    "import time,random,datetime\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib_venn\n",
    "import scanpy as sc\n",
    "import warnings\n",
    "import csv\n",
    "%matplotlib inline\n",
    "%load_ext memory_profiler\n",
    "\n",
    "# settings\n",
    "plt.rc('font', size = 8)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "sc.set_figure_params(dpi=300,dpi_save=600,\n",
    "                     frameon=False,\n",
    "                     fontsize=8)\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sc.settings.verbosity=2\n",
    "sc._settings.ScanpyConfig.n_jobs=-1\n",
    "\n",
    "# reproducibility\n",
    "rs = np.random.seed(42)\n",
    "\n",
    "# fps\n",
    "dfp = '/vast/palmer/pi/lim_janghoo/cl2292/'\n",
    "pfp = '/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/results/'\n",
    "pdfp = '/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/data/'\n",
    "sc.settings.figdir = pfp\n",
    "\n",
    "# load DEGs\n",
    "\n",
    "#OL-SCA1-cKI 30W\n",
    "\n",
    "dge = pd.read_csv('/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/results/250414_dge_SCA1-flwCre vs SCA1-flwoCre.csv')\n",
    "downsig_cKI = dge.loc[(dge['emd']<=-0.1)&(dge['pval_corrected']<0.01),:]\n",
    "upsig_cKI = dge.loc[(dge['emd']>=0.1)&(dge['pval_corrected']<0.01),:]\n",
    "\n",
    "#SCA1-KI 5-30W\n",
    "\n",
    "KI = pd.read_csv(pfp+'250414_dge_KI_imp.csv')\n",
    "downsig_KI = KI.loc[(KI['emd']<=-0.1)&(KI['pval_corrected']<0.01),:]\n",
    "upsig_KI = KI.loc[(KI['emd']>=0.1)&(KI['pval_corrected']<0.01),:]\n",
    "\n",
    "def filterdeg(df, ctype, timepoint=None):\n",
    "    if 'Cell type' not in df.columns or 'Gene' not in df.columns:\n",
    "        raise ValueError(\"DataFrame must contain 'Cell type' and 'Gene' columns.\")\n",
    "    \n",
    "    if timepoint is None:\n",
    "        filtered_df = df.loc[df['Cell type'] == ctype, :]\n",
    "    else:\n",
    "        if 'timepoint' not in df.columns:\n",
    "            raise ValueError(\"DataFrame must contain 'timepoint' column for filtering by timepoint.\")\n",
    "        filtered_df = df.loc[(df['Cell type'] == ctype) & (df['timepoint'] == timepoint), :]\n",
    "\n",
    "    return filtered_df['Gene'].to_list()\n",
    "\n",
    "c_list = downsig_cKI['Cell type'].unique().tolist()\n",
    "nrow = len(c_list)\n",
    "\n",
    "fig, axs = plt.subplots(nrow, 1, figsize=(2, nrow*2))\n",
    "for i, c in enumerate(c_list): \n",
    "    # Get sets of genes for venn diagrams\n",
    "    cKIdown = set(filterdeg(downsig_cKI, ctype=c))\n",
    "    KIdown = set(filterdeg(downsig_KI, ctype=c))\n",
    "\n",
    "    # Plot the Venn diagram for \"DOWN\"\n",
    "    ax = axs[i]  # Select the corresponding subplot\n",
    "    matplotlib_venn.venn2([cKIdown, KIdown], ('OL-SCA1-cKI 30W DOWN', 'SCA1-KI 5-30W DOWN'),\n",
    "                          set_colors=('#FF9999','#9999FF'), alpha = 0.8, ax=ax)\n",
    "    matplotlib_venn.venn2_circles([cKIdown, KIdown], linewidth=0.6, ax=ax)\n",
    "    ax.set_title(f\"{c} DOWN\")\n",
    "    \n",
    "fig.savefig(os.path.join(pfp, '250415_venn_overlappaing DEG_down.pdf'), dpi=300, bbox_inches='tight')\n",
    "\n",
    "fig, axs = plt.subplots(nrow, 1, figsize=(2, nrow*2))\n",
    "for i, c in enumerate(c_list): \n",
    "    # Get sets of genes for venn diagrams\n",
    "    cKIup = set(filterdeg(upsig_cKI, ctype=c))\n",
    "    KIup = set(filterdeg(upsig_KI, ctype=c))\n",
    "\n",
    "    # Plot the Venn diagram for \"UP\"\n",
    "    ax = axs[i]  # Select the corresponding subplot\n",
    "    matplotlib_venn.venn2([cKIup, KIup], ('OL-SCA1-cKI 30W UP', 'SCA1-KI 5-30W UP'),\n",
    "                          set_colors=('#FF9999','#9999FF'), alpha = 0.8, ax=ax)\n",
    "    matplotlib_venn.venn2_circles([cKIup, KIup], linewidth=0.6, ax=ax)\n",
    "    ax.set_title(f\"{c} UP\")\n",
    "    \n",
    "fig.savefig(os.path.join(pfp, '250415_venn_overlappaing DEG_up.pdf'), dpi=300, bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
