{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93438554",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, glob, re, math, pickle\n",
    "import scprep, magic, phate\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import sparse as sp\n",
    "import time,random,datetime\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "from typing import Dict, Optional\n",
    "import tables\n",
    "import seaborn as sns\n",
    "from bbknn import bbknn\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "%load_ext memory_profiler\n",
    "\n",
    "# reproducibility\n",
    "rs = np.random.seed(42)\n",
    "\n",
    "# fps\n",
    "dfp = '/gpfs/gibbs/pi/lim_janghoo/cl2292/'\n",
    "pfp = '/gpfs/gibbs/pi/lim_janghoo/cl2292/SCA1_OL/results/'\n",
    "pdfp = '/gpfs/gibbs/pi/lim_janghoo/cl2292/SCA1_OL/data/'\n",
    "sc.settings.figdir = pfp\n",
    "\n",
    "# settings\n",
    "plt.rc('font', size = 8)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "sc.set_figure_params(dpi=300,dpi_save=600,\n",
    "                     frameon=False,\n",
    "                     fontsize=8)\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sc.settings.verbosity=2\n",
    "sc._settings.ScanpyConfig.n_jobs=-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1edec158",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='250414_OL-SCA1-cKI_imp.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit adata = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4426bf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[(adata.obs['ctype']=='OPC')|(adata.obs['ctype']=='OL'),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2c93119",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X=adata.layers[\"counts\"]\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.sqrt(adata,chunked=True,chunk_size=10000)\n",
    "adata.raw = adata\n",
    "\n",
    "\n",
    "# calc embeddings for batch corrected\n",
    "start = time.time()\n",
    "print('starting embeddings...')\n",
    "sc.tl.pca(adata,n_comps=100)\n",
    "#sc.external.pp.bbknn(adata,batch_key='batch')\n",
    "bbknn(adata,batch_key='batch') #pip install bbknn\n",
    "sc.tl.leiden(adata,resolution=1) \n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color = ['genotype','leiden'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2231e58f",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = {\n",
    "    'OPC':['Pdgfra','Cspg4'],\n",
    "    'OPC1':['Pcdh15'], #,'Ptprz1'\n",
    "    'OPC2':['Caln1'], #'Car10'\n",
    "    'COP':['Fyn','Tcf7l2'], #'Bmp4','Neu4'\n",
    "    'NFOL':['Man1a','Synpr'], #,'Prom1','Tspan2' #marques 2018, Matson 2022; Hilscher2022 'Enpp6','Mpzl1','Itpr2'  \n",
    "    'OL':['Plp1','Mag','Mog','Mbp','Mobp','Opalin'], #,'Hapln2','Dpy19l1'\n",
    "    'MOL1':['Egr2'], #'Kirrel3','Egr2',\n",
    "    'MOL2': ['Klk6',], #'Cyp27a1','Rab37'\n",
    "    'MOL4': ['Ttr',],\n",
    "    'MOL5/6' : ['Ptgds'], #,'Grm3','Car2'\n",
    "    'OL3':['Fgf14','Rbfox1','Nrxn1']}\n",
    "\n",
    "#sc.tl.dendrogram(adata, groupby = 'ctype')\n",
    "sc.pl.dotplot(adata, markers, groupby = 'leiden', standard_scale='var', use_raw=False, dendrogram=True,\n",
    "#              save = '_241025_KI_OLsubtype.pdf'\n",
    "             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c41d651",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate cell type\n",
    "opc1 = ['5']\n",
    "opc2 = ['6']\n",
    "\n",
    "cop = ['8']\n",
    "nfol = ['9']\n",
    "\n",
    "ol1 = ['1','2']\n",
    "ol2 = ['0']\n",
    "ol3 = ['3','4','7']\n",
    "\n",
    "ctype = []\n",
    "for i in adata.obs['leiden'] : # verbose loop for quality-assurance\n",
    "    if i in opc1 :\n",
    "        ctype.append('OPC1') \n",
    "    elif i in opc2 :\n",
    "        ctype.append('OPC2')\n",
    "    elif i in cop :\n",
    "        ctype.append('COP')\n",
    "    elif i in nfol :\n",
    "        ctype.append('NFOL')\n",
    "    elif i in ol1 :\n",
    "        ctype.append('OL1')\n",
    "    elif i in ol2 :\n",
    "        ctype.append('OL2')\n",
    "    elif i in ol3 :\n",
    "        ctype.append('OL3')\n",
    "    else:\n",
    "        raise ValueError('Encountered unclassifiable cell type for sample {}'.format(i))\n",
    "adata.obs['ctype']=ctype\n",
    "sc.pl.umap(adata, color = ['ctype'])\n",
    "\n",
    "cmap_ctype={'OPC1': '#e41a1c',\n",
    "              'OPC2':'#377eb8',\n",
    "              'COP':'#4daf4a',\n",
    "              'NFOL':'#f781bf',\n",
    "              'OL1':'#984ea3',\n",
    "              'OL2':'#ff7f00',\n",
    "              'OL3':'#a65628'}\n",
    "\n",
    "adata.uns['ctype_colors']=np.array(list(cmap_ctype.values()),dtype=object)\n",
    "adata.obs['ctype']=adata.obs['ctype'].cat.reorder_categories(new_categories=('OPC1','OPC2','COP','NFOL','OL1','OL2','OL3'),\n",
    "                                         ordered=True)\n",
    "sc.pl.umap(adata, color = ['ctype'])\n",
    "\n",
    "cmap_genotype={'SCA1-fl/+': '#4683B5',\n",
    "            'SCA1-fl/NG2-Cre':'#FFA600'}\n",
    "\n",
    "adata.uns['genotype_colors']=np.array(list(cmap_genotype.values()),dtype=object)\n",
    "adata.obs['genotype']=adata.obs['genotype'].cat.reorder_categories(new_categories=list(cmap_genotype.keys()),\n",
    "                                         ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb63cf09",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_indices = np.random.permutation(list(range(adata.shape[0])))\n",
    "size = 40\n",
    "\n",
    "sc.pl.umap(adata[random_indices, :], color = 'genotype', size = size,\n",
    "           save = '_250414_OL-SCA-cKI_OLsub_genotype.pdf'\n",
    "          )\n",
    "sc.pl.umap(adata[random_indices, :], color = 'ctype', size = size,\n",
    "           save = '_240414_OL-SCA-cKI_OLsub_ctype.pdf'\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61d11dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_counts = adata.obs.groupby(['ctype','genotype','batch']).size().reset_index(name='cell_count')\n",
    "total_counts = adata.obs.groupby(['genotype','batch']).size().reset_index(name='total_cells')\n",
    "cell_counts = cell_counts.merge(total_counts, on=['genotype','batch'])\n",
    "cell_counts['proportion(%)'] = cell_counts['cell_count']*100 / cell_counts['total_cells']\n",
    "cell_counts = cell_counts.sort_values(by=['ctype', 'genotype','batch'])\n",
    "cell_counts\n",
    "cell_counts.to_csv(os.path.join(pfp,'250415_OL-SCA1-cKI_OL sub proportion.csv'),index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d351271",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(os.path.join(pdfp,'250415_OL-SCA1-cKI_OLsub.h5ad'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa0b79d2",
   "metadata": {},
   "source": [
    "# Imputation and DGE analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e644fb27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import os, sys, glob, re, math, pickle\n",
    "import phate, scprep, magic\n",
    "import graphtools as gt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import time,random,datetime\n",
    "import networkx as nx\n",
    "from sklearn import metrics\n",
    "from sklearn import model_selection\n",
    "from scipy import sparse\n",
    "from scipy.stats import mannwhitneyu, tiecorrect, rankdata\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import scanpy as sc\n",
    "from sklearn.dummy import DummyClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.cluster import SpectralClustering, OPTICS, cluster_optics_dbscan, AgglomerativeClustering\n",
    "from bbknn import bbknn\n",
    "import warnings\n",
    "%matplotlib inline\n",
    "%load_ext memory_profiler\n",
    "import logging\n",
    "\n",
    "# settings\n",
    "plt.rc('font', size = 8)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "sc.set_figure_params(dpi=300,dpi_save=600,\n",
    "                     frameon=False,\n",
    "                     fontsize=8)\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sc.settings.verbosity=2\n",
    "sc._settings.ScanpyConfig.n_jobs=-1\n",
    "\n",
    "# reproducibility\n",
    "rs = np.random.seed(42)\n",
    "\n",
    "# utils\n",
    "def mwu(X,Y,gene_names,correction=None,debug=False) :\n",
    "    '''\n",
    "    Benjamini-Hochberg correction implemented. Can change to Bonferonni\n",
    "    gene_names (list)\n",
    "    if X,Y single gene expression array, input x.reshape(-1,1), y.reshape(-1,1)\n",
    "    NOTE: get zeros sometimes because difference (p-value is so small)\n",
    "    '''\n",
    "    p=pd.DataFrame()\n",
    "    print('Mann-Whitney U w/Benjamini/Hochberg correction\\n')\n",
    "    start = time.time()\n",
    "    for i,g in enumerate(gene_names) :\n",
    "        if i==np.round(np.quantile(np.arange(len(gene_names)),0.25)) :\n",
    "            print('... 25% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.5)) :\n",
    "            print('... 50% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.75)) :\n",
    "            print('... 75% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        p.loc[i,'Gene']=g\n",
    "        if (tiecorrect(rankdata(np.concatenate((np.asarray(X[:,i]),np.asarray(Y[:,i])))))==0) :\n",
    "            if debug :\n",
    "                print('P-value not calculable for {}'.format(g))\n",
    "            p.loc[i,'pval']=np.nan\n",
    "        else :\n",
    "            _,p.loc[i,'pval']=mannwhitneyu(X[:,i],Y[:,i]) # continuity correction is True\n",
    "    print('... mwu computed in {:.2f}-s\\n'.format(time.time() - start))\n",
    "    # ignore NaNs, since can't do a comparison on these (change numbers for correction)\n",
    "    p_corrected = p.loc[p['pval'].notna(),:]\n",
    "    new_pvals = multipletests(p_corrected['pval'],method='fdr_bh')\n",
    "    p_corrected['pval_corrected'] = new_pvals[1]\n",
    "    return p_corrected\n",
    "\n",
    "def log2aveFC(X,Y,gene_names,AnnData=None) :\n",
    "    '''not sensitivity to directionality due to subtraction\n",
    "    X and Y full arrays, subsetting performed here\n",
    "    `gene_names` (list): reduced list of genes to calc\n",
    "    `adata` (sc.AnnData): to calculate reduced list. NOTE: assumes X,Y drawn from adata.var_names\n",
    "    '''\n",
    "    if not AnnData is None :\n",
    "        g_idx = [i for i,g in enumerate(AnnData.var_names) if g in gene_names]\n",
    "        fc=pd.DataFrame({'Gene':AnnData.var_names[g_idx],\n",
    "                         'log2FC':np.log2(X[:,g_idx].mean(axis=0)) - np.log2(Y[:,g_idx].mean(axis=0))}) # returns NaN if negative value \n",
    "    else :\n",
    "        fc=pd.DataFrame({'Gene':gene_names,\n",
    "                         'log2FC':np.log2(X.mean(axis=0)) - np.log2(Y.mean(axis=0))})\n",
    "    return fc\n",
    "\n",
    "\n",
    "# fps\n",
    "dfp = '/vast/palmer/pi/lim_janghoo/cl2292/'\n",
    "pfp = '/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/results/'\n",
    "pdfp = '/vast/palmer/pi/lim_janghoo/cl2292/SCA1_OL/data/'\n",
    "sc.settings.figdir = pfp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c8969be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Imputation by genotype & timepoint\n",
    "\n",
    "wt30 = adata[(adata.obs['genotype']=='SCA1-fl/+'), :]\n",
    "mut30 = adata[(adata.obs['genotype']=='SCA1-fl/NG2-Cre'), :]\n",
    "\n",
    "#k=45, t=3\n",
    "print('Starting imputation for {}\\n'.format('30wk SCA1-fl/+'))\n",
    "tic = time.time()\n",
    "\n",
    "\n",
    "wt30.obs['value'] = 0\n",
    "sc.pp.pca(wt30)\n",
    "sc.pp.neighbors(wt30, n_pcs=45)\n",
    "\n",
    "\n",
    "# MAGIC\n",
    "G = gt.Graph(data=wt30.obsp['connectivities']+sparse.diags([1]*wt30.shape[0],format='csr'),\n",
    "             precomputed='adjacency',\n",
    "             use_pygsp=True)\n",
    "G.knn_max = None\n",
    "\n",
    "magic_op=magic.MAGIC().fit(X=wt30.X,graph=G) # running fit_transform produces wrong shape\n",
    "wt30.layers['imputed']=magic_op.transform(wt30.X,genes='all_genes')\n",
    "\n",
    "print('\\n  imputation in {:.2f}-min'.format((time.time() - tic)/60))\n",
    "\n",
    "\n",
    "print('\\n Starting imputation for {}\\n'.format('30wk SCA1-fl/NG2-Cre'))\n",
    "tic = time.time()\n",
    "\n",
    "mut30.obs['value'] = 0\n",
    "sc.pp.pca(mut30)\n",
    "sc.pp.neighbors(mut30, n_pcs=45)\n",
    "\n",
    "# MAGIC\n",
    "G = gt.Graph(data=mut30.obsp['connectivities']+sparse.diags([1]*mut30.shape[0],format='csr'),\n",
    "             precomputed='adjacency',\n",
    "             use_pygsp=True)\n",
    "G.knn_max = None\n",
    "\n",
    "magic_op=magic.MAGIC().fit(X=mut30.X,graph=G) # running fit_transform produces wrong shape\n",
    "mut30.layers['imputed']=magic_op.transform(mut30.X,genes='all_genes')\n",
    "\n",
    "print('\\n  imputation in {:.2f}-min'.format((time.time() - tic)/60))\n",
    "\n",
    "adata =wt30.concatenate([mut30] ,batch_key='concat')\n",
    "\n",
    "# save data objects\n",
    "adata.write(os.path.join(pdfp,'250415_OL-SCA1-cKI_OLsub_imp.h5ad'))\n",
    "print('saved @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f826daa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EMD; IMP\n",
    "wt = adata[(adata.obs['genotype']=='SCA1-fl/+'), :]\n",
    "mut = adata[(adata.obs['genotype']=='SCA1-fl/NG2-Cre'), :]\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='ctype'\n",
    "    fname = 'OL-SCA1-cKI OL subtype' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['30wk'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in wt.obs[group].unique():\n",
    "            start = time.time()\n",
    "            print('\\n{}, SCA1 vs WT'.format(i))\n",
    "            print('----')\n",
    "            X = wt[(wt.obs[group]==i)&(wt.obs['timepoint']==t), :].layers['imputed']\n",
    "            Y = mut[(mut.obs[group]==i)&(mut.obs['timepoint']==t), :].layers['imputed']\n",
    "            \n",
    "\n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(Y,X,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(Y,X, wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = pd.concat([dge,dt], ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dge.to_csv(os.path.join(pfp,'250415_dge_'+fname+'.csv'),index=False)\n",
    "#     dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "#     dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "#     dgeup.to_csv(os.path.join(pfp,'240723_dge_'+fname+'_SCA1flwCre_up.csv'),index=False)\n",
    "#     dgedown.to_csv(os.path.join(pfp,'240723_dge_'+fname+'_SCA1flwCre_down.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
